### DISCUSSION TITLE
fail in 3D meshing

### DISCUSSION THREAD

**Discussion by Yuanguo1994** [CsatiZoltan/CristalX#74] [Q&A] [answered]
Hi, CsatiZoltan

After the microstructure building, I inputed it to the Salome following the steps you wrote in the documentation. The mesh for 2D microstrucuture was successful, but the when I extruded it 3D model, some errors always come like this:

![image](https://user-images.githubusercontent.com/79046436/124399358-c0e03100-dd12-11eb-8bac-004baffa0e21.png)

I searhced for the errors in the google web, some think it is because of the patition. So I made the patition for the model, but the same error came out again.  Do you know what cause the fail in meshing? and could you please give me some suggestion or tips on solving the problem? The model picture is shown following:
![image](https://user-images.githubusercontent.com/79046436/124399453-832fd800-dd13-11eb-95f0-d1ff9b3f599f.png)

**Comment by CsatiZoltan** [comment, marked as answer]
Hi @Yuanguo1994 

It is nice to see that you progress well in going from an image to a 3D geometry. The error messages you show are well-known to me. As I mentioned in a [previous post](https://github.com/CsatiZoltan/CristalX/discussions/68#discussioncomment-945021), I will make a step by step tutorial soon, but I am quite busy these days. Nevertheless, I try to help you with this specific problem.

The error I see on the screenshot is typically caused by two issues:
1. Let's take two neighboring grains that share a common face (line in 2D, surface in 3D). You want that face to be shared by the two incident grains. In reality, Salome considers that face as two separate overlapping faces, one belonging to grain 1 and the other belonging to grain 2. Therefore, when you create a mesh, the meshes will be independent on the two overlapping faces, i.e. you get a non-conforming mesh. The solution is indeed what you wrote: inform Salome that the common face is unique. This is called [partitioning](https://docs.salome-platform.org/latest/gui/GEOM/partition_explanation.html). In my experience, if you do partitioning on the 2D geometry, it does not transfer to the extruded (3D) geometry, which makes sense as new objects are created. Therefore, you must perform the partitioning on the final (i.e. the extruded) geometry by clicking on *Operations* --> *Partition*. In the *Partition* frame, choose your extruded geometry as *Objects* and leave the *Tool Objects* empty. The *Resulting Type* should be "Solid". You can check the resulting partitioned geometry for possible errors from the *Inspection* --> *Detect Self-intersections* menu. If you get self-intersecting faces, you can export them to an object and see where the overlapping faces are. One way to avoid self-intersecting faces is by increasing the tolerance in the *Repair* --> *Limit tolerance* menu.
2. In some cases, even if you can achieve a non-overlapping geometry described in the previous point, you still get meshing errors. This is so because your geometry is complicated (locally high curvature, acute angles, etc.). Even excellent mesh generators like Netgen can fail to create a coarse conforming mesh in that case. You have several choices then.
  a. Create a finer mesh.
  b. Try alternative mesh generators (in Salome, Netgen is the best among the free ones, but it's worth trying others).
  c. Try to first create a surface mesh, check if it is non-overlapping, and then create the volumetric mesh in the second step.
  d. Simplify the geometry.
  When you receive meshing errors, you can publish the problematic sub-shapes into your study. This way, you can check more closely what happened, and choose a strategy accordingly.

Meshing complicated geometries is not straightforward. Most often you cannot automate it as it requires manual work and testing various options.

**Reply by Yuanguo1994** [reply]
Thanks very much, it seems the mesh is successful followin your tips, see the picture below. Another question is that how can I export it to gmesh format? I havent found any tutorial about that? Could you plz tell me how to do that? or recommend some references? I want to use the mesh to do some simple simulations in software like abaqus or fepx. 
![image](https://user-images.githubusercontent.com/79046436/124481884-ab194d00-dda0-11eb-9e75-ce60038c0c6d.png)

**Reply by CsatiZoltan** [reply]
For curiosity, how did you solve your problem about meshing?

> how can I export it to gmesh format?

Select your mesh (`Mesh_4` as I see on the screenshot), right click -- > *Export* --> *MED file*. You can read your .med file in Gmsh. Then you can export it to .msh format from within Gmsh.

**Reply by Yuanguo1994** [reply]
I just increased the 'limit tolerance' to about 1e5, and then it works. The material I work on is directionally solidified material consist of columnar grains. I am  looking for a way to build the microstructure for a long time, so really thank you for providing the method. By the way, I am working on the fatigue behavior of the material,  mainly focus on experiments work, not very familiar with the simulation or model building work. Sometimes, I might ask some very simple or 'fool' questions. Sorry for that.

**Reply by CsatiZoltan** [reply]
> I just increased the 'limit tolerance' to about 1e5, and then it works.

Be careful to choose reasonable values. If the tolerance is too large compared to the geometrical features (depends on the unit they are given), you lose precision. Unless you scaled the geometry, the unit is pixel (as it comes from an image). So even the largest value in the geometry (let alone the small grains) are much smaller than 1e5, meaning that you suffer a huge loss of precision.

> The material I work on is directionally solidified material consist of columnar grains.

Then I understand why you want to extrude the geometry.

>  mainly focus on experiments work, not very familiar with the simulation or model building work. Sometimes, I might ask some very simple or 'fool' questions.

No problem. On the contrary, my work is completely numerical, so I know nothing about experiments!

If I could provide a satisfying answer to a topic, mark it as accepted and feel free to open a new discussion or return to an existing topic later in case you have further problems.
